### PR TITLE
[controller] the interface of the task queue for controller

### PR DESCRIPTION
**Pull request by newpoo**
This is the main interface for the task queue library.

Haven't added implementation.

Basically, the worker needs to periodically call resetZombieTasks() and removeFinishedTasks() to do house cleaning.

### CODE REVIEW

**Comment by angxu** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[controller/task-queue/src/main/java/Task.java, line 31 at r1](https://reviewable.io:443/reviews/pinterest/rocksplicator/22#-KahmzKLHpyYJyO6_BTl:-KahhsdOLeq8_vYTOITV:b-yowj21) ([raw file](https://github.com/pinterest/rocksplicator/blob/bbd50bc466d31178655a4bc15a8e0ed2c520489a/controller/task-queue/src/main/java/Task.java#L31)):*
> ```Java
>   public String clusterName;
>   public String body;
>   public Timestamp createdAt;
> ```

I'm not familiar with Timestamp. But I'd suggest to use primitive type(ie. long) in POJO because it can interact seamlessly with different data sources. Timestamp, on the other hand, can have serialization/deserialization problems.

---

*[controller/task-queue/src/main/java/TaskQueue.java, line 60 at r1](https://reviewable.io:443/reviews/pinterest/rocksplicator/22#-KahkLMwdOaHNjnVXCQp:-KahkLMwdOaHNjnVXCQq:b-7kp8vj) ([raw file](https://github.com/pinterest/rocksplicator/blob/bbd50bc466d31178655a4bc15a8e0ed2c520489a/controller/task-queue/src/main/java/TaskQueue.java#L60)):*
> ```Java
>    * @return the dequeued task, or null if no eligible task found
>    */
>   public Task dequeueTask(final String worker) {
> ```

Is it a blocking call? If so, please document it and add a timeout argument.

---

*[controller/task-queue/src/main/java/TaskQueue.java, line 176 at r1](https://reviewable.io:443/reviews/pinterest/rocksplicator/22#-KahjjSkuJTVXGEXiju5:-KahjjSlIJ-dTM69efiY:bp4z8t2) ([raw file](https://github.com/pinterest/rocksplicator/blob/bbd50bc466d31178655a4bc15a8e0ed2c520489a/controller/task-queue/src/main/java/TaskQueue.java#L176)):*
> ```Java
>    */
>   public List<Task> peekTasks(final String clusterName,
>                               final Integer state) {
> ```

Use primitive int.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pinterest/rocksplicator/22)*

**Comment by newpoo** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[controller/task-queue/src/main/java/TaskQueue.java, line 60 at r1](https://reviewable.io:443/reviews/pinterest/rocksplicator/22#-KahkLMwdOaHNjnVXCQp:-KahpdpP7Y_s2mA96iRV:b-yzhbko) ([raw file](https://github.com/pinterest/rocksplicator/blob/bbd50bc466d31178655a4bc15a8e0ed2c520489a/controller/task-queue/src/main/java/TaskQueue.java#L60)):*
<details><summary><i>Previously, angxu wrote…</i></summary><blockquote>

Is it a blocking call? If so, please document it and add a timeout argument.
</blockquote></details>

It isn't blocking. It will return "null" if there is no eligible task found or errors.

---

*[controller/task-queue/src/main/java/TaskQueue.java, line 176 at r1](https://reviewable.io:443/reviews/pinterest/rocksplicator/22#-KahjjSkuJTVXGEXiju5:-Kahq-4SsqYDqzkVF3Qd:bpqtrtl) ([raw file](https://github.com/pinterest/rocksplicator/blob/bbd50bc466d31178655a4bc15a8e0ed2c520489a/controller/task-queue/src/main/java/TaskQueue.java#L176)):*
<details><summary><i>Previously, angxu wrote…</i></summary><blockquote>

Use primitive int.
</blockquote></details>

As described in the comment, I want to use null to indicate no specified state is requested, i.e., return tasks in any state.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pinterest/rocksplicator/22)*

File: controller/task-queue/src/main/java/Task.java [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 package com.pinterest.rocksplicator.controller;
 
-import java.security.Timestamp;
```
Comment (newpoo) [inline controller/task-queue/src/main/java/Task.java]: Didn't notice the IDE imported this Timestamp.

**Comment by newpoo** [comment]
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[controller/task-queue/src/main/java/Task.java, line 31 at r1](https://reviewable.io:443/reviews/pinterest/rocksplicator/22#-KahmzKLHpyYJyO6_BTl:-KaiIuTJ9IOT-XNVkJa9:b-sjrham) ([raw file](https://github.com/pinterest/rocksplicator/blob/bbd50bc466d31178655a4bc15a8e0ed2c520489a/controller/task-queue/src/main/java/Task.java#L31)):*
<details><summary><i>Previously, angxu wrote…</i></summary><blockquote>

I'm not familiar with Timestamp. But I'd suggest to use primitive type(ie. long) in POJO because it can interact seamlessly with different data sources. Timestamp, on the other hand, can have serialization/deserialization problems.
</blockquote></details>

These fields of Timestamp type won't be serialized into JSON blob. They are standalone field in MySQL of type DataTime. Since JDBC has good support for translating between MySQL DateTime and java.sql.Timestamp, and Timestamp has some helper functions which are useful for us, I'd keep them as Timestamp unless you have a concrete example where long is better than Timestamp in our use case here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pinterest/rocksplicator/22)*

**Comment by angxu** [comment]
:lgtm:

---

Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[controller/task-queue/src/main/java/Task.java, line 31 at r1](https://reviewable.io:443/reviews/pinterest/rocksplicator/22#-KahmzKLHpyYJyO6_BTl:-KaiJy8n84e9sqgVFkXt:b-q0pxau) ([raw file](https://github.com/pinterest/rocksplicator/blob/bbd50bc466d31178655a4bc15a8e0ed2c520489a/controller/task-queue/src/main/java/Task.java#L31)):*
<details><summary><i>Previously, newpoo (Bo Liu) wrote…</i></summary><blockquote>

These fields of Timestamp type won't be serialized into JSON blob. They are standalone field in MySQL of type DataTime. Since JDBC has good support for translating between MySQL DateTime and java.sql.Timestamp, and Timestamp has some helper functions which are useful for us, I'd keep them as Timestamp unless you have a concrete example where long is better than Timestamp in our use case here.
</blockquote></details>

okay, I'm fine with it if it's supported by jdbc library.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pinterest/rocksplicator/22#-:-KaiKLLyb2QeIS9pwCBF:bnfp4nl)*

**Comment by angxu** [comment]
Reviewed 1 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pinterest/rocksplicator/22)*

**Comment by shuzhang1989** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pinterest/rocksplicator/22#-:-KaiVpqvlY1q35NDYQR6:bnfp4nl)*

**Pull request merged**